src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+
+// This membership cannot be erased (either it is not expired or not in grace period and/or not the owner)
+error CantEraseMembership(uint256 idCommitment);
+
+contract Membership {
+    using SafeERC20 for IERC20;
+
+    /// @notice Address of the Price Calculator used to calculate the price of a new membership
+    IPriceCalculator public priceCalculator;
+
+    /// @notice Maximum total rate limit of all memberships in the tree
+    uint32 public maxTotalRateLimitPerEpoch;
+
+    /// @notice Maximum rate limit of one membership
+    uint32 public maxRateLimitPerMembership;
+
+    /// @notice Minimum rate limit of one membership
+    uint32 public minRateLimitPerMembership;
+
+    /// @notice Membership billing period
+    uint32 public billingPeriod;
+
+    /// @notice Membership grace period
+    uint32 public gracePeriod;
+
+    /// @notice balances available to withdraw
+    mapping(address => mapping(address => uint256)) public balancesToWithdraw; // holder ->  token -> balance
+
+    /// @notice Total rate limit of all memberships in the tree
+    uint256 public totalRateLimitPerEpoch;
+
+    /// @notice List of registered memberships (IDCommitment to membership metadata)
+    mapping(uint256 => MembershipInfo) public members;
+
+    /// @notice The index of the next member to be registered
+    uint32 public commitmentIndex;
+
+    /// @notice track available indices that are available due to expired memberships being removed
+    uint32[] public availableExpiredIndices;
+
+    /// @dev Oldest membership
+    uint256 public head = 0;
+
+    /// @dev Newest membership
+    uint256 public tail = 0;
+
+    struct MembershipInfo {
+        /// @notice idCommitment of the previous membership
+        uint256 prev;
+        /// @notice idCommitment of the next membership
+        uint256 next;
+        /// @notice amount of the token used to acquire this membership
+        uint256 amount;
+        /// @notice numPeriods
+        uint32 numberOfPeriods;
+        /// @notice timestamp of when the grace period starts for this membership
+        uint256 gracePeriodStartDate;
+        /// @notice duration of the grace period
+        uint32 gracePeriod;
+        /// @notice the user message limit of each member
+        uint32 userMessageLimit;
+        /// @notice the index of the member in the set
+        uint32 index;
+        /// @notice address of the owner of this membership
+        address holder;
+        /// @notice token used to acquire this membership
+        address token;
+    }
+
+    /// @notice Emitted when a membership is erased due to having exceeded the grace period or the owner having chosen
+    /// to not extend it
+    /// @param idCommitment the idCommitment of the member
+    /// @param userMessageLimit the rate limit of this membership
+    /// @param index the index of the membership in the merkle tree
+    event MemberExpired(uint256 idCommitment, uint32 userMessageLimit, uint32 index);
```

Comment:
For nwaku, it would only require the index, right? I added the idCommitment and userMessageLimit in case we want to calculate the rateCommitment on the nwaku side, but is it really necessary?

Reply:
index should be enough but not sure if the fact that memberships expire may require idcommit/msglimit. I mean, I don't think it would be strictly necessary but nice for checking that the index you are removing corresponds to the commitment you are expecting. But yeah, a nice to have, not required.